File: app/src/main/java/info/nightscout/androidaps/plugins/PumpInsightLocal/app_layer/remote_control/SetDateTimeMessage.java
package info.nightscout.androidaps.plugins.PumpInsightLocal.app_layer.remote_control;

import info.nightscout.androidaps.plugins.PumpInsightLocal.app_layer.AppLayerMessage;
import info.nightscout.androidaps.plugins.PumpInsightLocal.app_layer.Service;
import info.nightscout.androidaps.plugins.PumpInsightLocal.descriptors.MessagePriority;
import info.nightscout.androidaps.plugins.PumpInsightLocal.descriptors.PumpTime;
import info.nightscout.androidaps.plugins.PumpInsightLocal.utils.ByteBuf;

public class SetDateTimeMessage extends AppLayerMessage {

    private PumpTime pumpTime;

    public SetDateTimeMessage() {
        super(MessagePriority.NORMAL, false, true, Service.CONFIGURATION);
    }

    @Override
    protected ByteBuf getData() {
        ByteBuf byteBuf = new ByteBuf(7);
        byteBuf.putUInt16LE(pumpTime.getYear());
        byteBuf.putUInt8((short) pumpTime.getMonth());
        byteBuf.putUInt8((short) pumpTime.getDay());
        byteBuf.putUInt8((short) pumpTime.getHour());
        byteBuf.putUInt8((short) pumpTime.getMinute());
        byteBuf.putUInt8((short) pumpTime.getSecond());
        return byteBuf;
    }

    public void setPumpTime(PumpTime pumpTime) {
        this.pumpTime = pumpTime;
    }
}

File: pump/insight/src/main/java/info/nightscout/androidaps/plugins/pump/insight/app_layer/remote_control/SetDateTimeMessage.kt
package info.nightscout.androidaps.plugins.pump.insight.app_layer.remote_control

import info.nightscout.androidaps.plugins.pump.insight.app_layer.AppLayerMessage
import info.nightscout.androidaps.plugins.pump.insight.app_layer.Service
import info.nightscout.androidaps.plugins.pump.insight.descriptors.MessagePriority
import info.nightscout.androidaps.plugins.pump.insight.descriptors.PumpTime
import info.nightscout.androidaps.plugins.pump.insight.utils.ByteBuf

class SetDateTimeMessage : AppLayerMessage(MessagePriority.NORMAL, false, true, Service.CONFIGURATION) {

    internal lateinit var pumpTime: PumpTime
    override val data: ByteBuf
        get() {
            val byteBuf = ByteBuf(7).apply {
                putUInt16LE(pumpTime.year)
                putUInt8(pumpTime.month.toShort())
                putUInt8(pumpTime.day.toShort())
                putUInt8(pumpTime.hour.toShort())
                putUInt8(pumpTime.minute.toShort())
                putUInt8(pumpTime.second.toShort())
            }
            return byteBuf
        }
}